pkg/cloudprovider/provider/vmware-cloud-director/provider.go

```diff
@@ -0,0 +1,531 @@
+}
+
+func (p *provider) MachineMetricsLabels(machine *clusterv1alpha1.Machine) (map[string]string, error) {
+	labels := make(map[string]string)
+
+	c, _, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err == nil {
+		labels["size"] = fmt.Sprintf("%d-cpus-%d-mb", c.CPUs, c.MemoryMB)
+		labels["vapp"] = c.VApp
+		labels["vdc"] = c.VDC
+		labels["organization"] = c.Organization
+	}
+
+	return labels, err
+}
+
+func (p *provider) MigrateUID(_ context.Context, _ *clusterv1alpha1.Machine, _ types.UID) error {
+	return nil
+}
+
+func (p *provider) SetMetricsForMachines(machines clusterv1alpha1.MachineList) error {
+	return nil
+}
+
+func (p *provider) Validate(_ context.Context, spec clusterv1alpha1.MachineSpec) error {
+	c, _, _, err := p.getConfig(spec.ProviderSpec)
+	if err != nil {
+		return fmt.Errorf("failed to parse config: %w", err)
+	}
+
+	client, err := NewClient(c.Username, c.Password, c.Organization, c.URL, c.VDC, c.AllowInsecure)
+	if err != nil {
+		return fmt.Errorf("failed to create VMware Cloud Director client: %w", err)
+	}
+
+	// Ensure that the organization, VDC, and vApp exists.
+	org, vdc, vapp, err := client.GetOrganizationVDCAndVapp(c.VApp)
+	if err != nil {
+		return err
+	}
+
+	// Ensure that the catalog exists.
+	catalog, err := org.GetCatalogByNameOrId(c.Catalog, true)
+	if err != nil {
+		return fmt.Errorf("failed to get catalog '%s': %w", c.Catalog, err)
+	}
+
+	// Ensure that the template exists in the catalog
+	// Catalog item can be a vApp template OVA or media ISO file.
+	catalogItem, err := catalog.GetCatalogItemByNameOrId(c.Template, true)
+	if err != nil {
+		return fmt.Errorf("failed to get template '%s' in catalog '%s': %w", c.Template, c.Catalog, err)
+	}
+	if c.DiskSizeGB != nil && catalogItem.CatalogItem.Size > *c.DiskSizeGB {
+		return fmt.Errorf("diskSizeGB '%v' cannot be less than the template size '%v': %w", *c.DiskSizeGB, catalogItem.CatalogItem.Size, err)
+	}
+
+	// Ensure that the network exists
+	// It can either be a vApp network or a vApp Org network.
+	_, err = GetVappNetworkType(c.Network, *vapp)
+	if err != nil {
+		return fmt.Errorf("failed to get network '%s' for vapp '%s': %w", c.Network, c.VApp, err)
+	}
+
+	if c.SizingPolicy != nil || c.PlacementPolicy != nil {
+		allPolicies, err := org.GetAllVdcComputePolicies(url.Values{})
+		if err != nil {
+			return fmt.Errorf("failed to get template all VDC compute policies: %w", err)
+		}
+
+		if c.SizingPolicy != nil && *c.SizingPolicy != "" {
+			sizingPolicy := getComputePolicy(*c.SizingPolicy, allPolicies)
+			if sizingPolicy == nil {
+				return fmt.Errorf("sizing policy '%s' doesn't exist", *c.SizingPolicy)
+			}
+		}
+
+		if c.PlacementPolicy != nil && *c.PlacementPolicy != "" {
+			placementPolicy := getComputePolicy(*c.PlacementPolicy, allPolicies)
+			if placementPolicy == nil {
+				return fmt.Errorf("placement policy '%s' doesn't exist", *c.SizingPolicy)
+			}
+		}
```

Comment:
As far as I see, we only validate if it is a valid/existing compute policy. What if someone provides a placement policy as a sizing policy (and vice-versa)? Does it matter at all?

Reply:
Unfortunately, there is no way to detect the type at the moment. That's why we just check for their existence. 

Also, it's eventually just a `ComputePolicy` so it doesn't really matter.


https://github.com/kubermatic/machine-controller/blob/vmware-cloud-director/pkg/cloudprovider/provider/vmware-cloud-director/helper.go#L89